text/0000-option-unwrap-unchecked.md

```diff
@@ -0,0 +1,73 @@
+- Feature Name: option_unwrap_unchecked
+- Start Date: 2015-4-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC adds a method to the `Option` enum which allows it to be
+unsafely unwrapped without checking for `None`.
+
+# Motivation
+
+In cases where a developer can determine that an `Option` is `Some(v)`
+but the compiler cannot, this method allows a developer to unsafely
+opt-in to an optimization which will avoid a conditional check. In my
+own uses I find cases where I know that an `Option` is `Some(v)`. I
+present a case from the `libcollections linked_list.rs`:
+
+    fn next(&mut self) -> Option<&'a A> {
+        if self.nelem == 0 {
+            return None;
+        }
+        self.head.as_ref().map(|head| {
+            self.nelem -= 1;
+            self.head = &head.next;
+            &head.value
+        })
+    }
+
+The check for `nelem == 0` already ensures that `self.head.as_ref()`
+cannot be `None` but the `map` method is unlikely to optimize out the
+check. Here, an unsafe method which could unwrap the `Option` would
+allow the compiler to optimize. I expect that this can be used in many
+more cases throughout both the standard library and other code. The
+outcome is likely a minor performance improvement, but one that should
+be possible for developers to make.
+
+# Detailed design
+
+I choose to name this method `unwrap_unchecked()` (commence
+bike-shedding). It is implemented quite simply by using the
+`unreachable` intrinsic to indicate that the `Option` cannot be `None`
+and then calling `unwrap()`. I expect that the compiler can them
+optimize out the conditional. A prototype implement can be found
```

Comment:
I've used the following to make an unchecked unwrap. I didn't get it to work with just unreachable in my case:

```rust
unsafe {
    ::std::intrinsics::assume(match next_opt {
        None => false,
        Some(_) => true,
    });
    $B = match next_opt {
        None => return None,
        Some(elt) => elt
    };
}
```

I think unreachable is preferable over assume, if it's enough possible.